graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java

```diff
@@ -216,85 +225,175 @@ public final void finish(VisitorContext visitorContext) {
+
+    private void generateNativeImageProperties(VisitorContext visitorContext) {
+        List<Map> json;
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectWriter writer = mapper.writer(new DefaultPrettyPrinter());
+
+        Optional<Path> projectDir = visitorContext.projectDir();
+
+        File userReflectJsonFile = projectDir
+                .map(projectPath -> Paths.get(projectPath.toString(), BASE_REFLECT_JSON).toFile())
+                .orElse(null);
+
+        if (userReflectJsonFile != null && userReflectJsonFile.exists()) {
+            try {
+                json = mapper.readValue(userReflectJsonFile, new TypeReference<List<Map>>() {
+                });
+            } catch (Throwable e) {
+                visitorContext.fail("Error parsing base reflect.json: " + BASE_REFLECT_JSON, null);
+                return;
+            }
+        } else {
+            json = new ArrayList<>();
+        }
+
+        if (CollectionUtils.isEmpty(classes) && CollectionUtils.isEmpty(arrays) && CollectionUtils.isEmpty(json)) {
+            return;
+        }
+
+        try {
+            String path = buildNativeImagePath();
+            String reflectFile = path + REFLECTION_CONFIG_JSON;
+            String propsFile = path + NATIVE_IMAGE_PROPERTIES;
+
+            visitorContext.visitMetaInfFile(propsFile).ifPresent(gf -> {
+                visitorContext.info("Writing " + NATIVE_IMAGE_PROPERTIES + " file to destination: " + gf.getName());
+                try (PrintWriter w = new PrintWriter(gf.openWriter())) {
+                    w.println("Args = -H:ReflectionConfigurationResources=${.}/reflection-config.json");
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            });
+            final Optional<GeneratedFile> generatedFile = visitorContext.visitMetaInfFile(reflectFile);
+            generatedFile.ifPresent(gf -> {
+                for (Map<String, Object> value : classes.values()) {
+                    json.add(value);
+                }
+
+                for (String array : arrays) {
+                    json.add(CollectionUtils.mapOf(
+                            NAME, "[L" + array.substring(0, array.length() - 2) + ";",
+                            ALL_DECLARED_CONSTRUCTORS, true
+                    ));
+                }
+
+                try (Writer w = gf.openWriter()) {
+                    visitorContext.info("Writing " + REFLECTION_CONFIG_JSON + " file to destination: " + gf.getName());
+
+                    writer.writeValue(w, json);
+                } catch (IOException e) {
+                    visitorContext.fail("Error writing " + REFLECTION_CONFIG_JSON + ": " + e.getMessage(), null);
+                }
+            });
+        } finally {
+            classes.clear();
+            arrays.clear();
+        }
+    }
+
+    private void generateResourceConfig(VisitorContext visitorContext) {
+        ObjectMapper mapper = new ObjectMapper();
```

Comment:
We should be using a shared object mapper instead of creating a new one for each operation